File: internal/core/workflow.go
package core

import "fmt"

type WorkflowStatus string

const (
	StatusActive   WorkflowStatus = "active"
	StatusDisabled WorkflowStatus = "disabled"
	StatusDraft    WorkflowStatus = "draft"
)

type TriggerType string

const (
	TriggerWebhook  TriggerType = "webhook"
	TriggerSchedule TriggerType = "schedule"
	TriggerManual   TriggerType = "manual"
)

type Trigger struct {
	Type   TriggerType            `json:"type" yaml:"type"`
	Config map[string]interface{} `json:"config" yaml:"config"`
}

type ScheduleConfig struct {
	Cron     string `json:"cron"`
	Timezone string `json:"timezone"` // Optional
}

type NextStep struct {
	StepID    string `json:"step_id" yaml:"step_id"`
	Condition string `json:"condition"`
}

type Step struct {
	ID         string                 `json:"id" yaml:"id"`
	Action     string                 `json:"action" yaml:"action"`
	Params     map[string]interface{} `json:"params" yaml:"params"`
	NextSteps  []NextStep             `json:"next_steps" yaml:"next_steps"`
	PassOutput bool                   `json:"pass_output" yaml:"pass_output"`
	Retries    int                    `json:"retries" yaml:"retries"`
	Timeout    int                    `json:"timeout" yaml:"timeout"` // Seconds
}

type Workflow struct {
	ID      string                 `json:"id" yaml:"id"`
	Name    string                 `json:"name" yaml:"name"`
	Status  WorkflowStatus         `json:"status" yaml:"status"`
	Vars    map[string]interface{} `json:"vars" yaml:"vars"`
	Trigger Trigger                `json:"trigger" yaml:"trigger"`
	Steps   []Step                 `json:"steps" yaml:"steps"`
	// StepsMap for quick lookup
	StepsMap map[string]*Step `json:"-"`
}

func (w *Workflow) Validate() error {
	// Check initial step exists
	initialStep := w.Trigger.Config["initial_step"].(string)
	if _, exists := w.StepsMap[initialStep]; !exists {
		return fmt.Errorf("initial step %s not found", initialStep)
	}

	// Validate all next_steps references
	for _, step := range w.Steps {
		for _, next := range step.NextSteps {
			if _, exists := w.StepsMap[next.StepID]; !exists {
				return fmt.Errorf("step %s references invalid next step: %s",
					step.ID, next.StepID)
			}
		}
	}

	return nil
}

File: ui/src/types/workflow.ts
/**
 * OwlFlow Workflow Schema and AST Interfaces
 */

export type WorkflowStatus = 'active' | 'disabled' | 'draft';

export type TriggerType = 'webhook' | 'schedule' | 'manual';

export interface WebhookTriggerConfig {
  initial_step: string;
  path?: string;
  secret?: string;
  [key: string]: any;
}

export interface ScheduleTriggerConfig {
  initial_step: string;
  cron: string;
  timezone?: string;
  [key: string]: any;
}

export interface ManualTriggerConfig {
  initial_step: string;
  description?: string;
  [key: string]: any;
}

export interface Trigger {
  type: TriggerType;
  config: WebhookTriggerConfig | ScheduleTriggerConfig | ManualTriggerConfig | Record<string, any>;
}

export interface NextStep {
  step_id: string;
  condition?: string;
}

export interface Step {
  id: string;
  action: string;
  params?: Record<string, any>;
  next_steps?: NextStep[];
  pass_output?: boolean;
  retries?: number;
  timeout?: number;
  [key: string]: any;
}

export interface Workflow {
  id: string;
  name: string;
  status: WorkflowStatus;
  vars?: Record<string, any>;
  trigger: Trigger;
  steps: Step[];
  [key: string]: any;
}

export type DiagnosticSeverity = 'error' | 'warning' | 'info';

export interface DiagnosticRange {
  startLine: number;
  startCol: number;
  endLine: number;
  endCol: number;
}

export interface Diagnostic {
  code: string;
  message: string;
  severity: DiagnosticSeverity;
  path?: (string | number)[];
  range?: DiagnosticRange;
  suggestion?: string;
}

export interface ValidationResult {
  isValid: boolean;
  errors: Diagnostic[];
  warnings: Diagnostic[];
  diagnostics: Diagnostic[];
  reachableStepIds: Set<string>;
  unreachableStepIds: Set<string>;
  cycles: string[][];
}

export interface ExecutionContext {
  trigger: {
    payload?: any;
    headers?: Record<string, string | string[] | any>;
    query?: Record<string, any>;
    time?: string;
    timezone?: string;
    type?: string;
    [key: string]: any;
  };
  steps: Record<string, { output: any; [key: string]: any }>;
  vars: Record<string, any>;
  parent: any[];
  // Compatibility aliases
  TriggerData?: any;
  inputs?: any;
  variables?: any;
  [key: string]: any;
}

export type StepExecutionStatus = 'pending' | 'running' | 'success' | 'failed' | 'bypassed' | 'skipped';

export interface StepExecution {
  stepId: string;
  action: string;
  status: 'success' | 'failed';
  startedAt: string;
  durationMs: number;
  resolvedParams: Record<string, any>;
  output: any;
  error?: string;
}

export interface TransitionExecution {
  fromStepId: string;
  toStepId: string;
  condition?: string;
  evaluatedResult: boolean;
  status: 'active' | 'bypassed' | 'error';
  error?: string;
}

export interface SimulationInput {
  type?: TriggerType;
  payload: Record<string, any>;
  headers?: Record<string, string | string[] | any>;
  query?: Record<string, any>;
  time?: string;
  timezone?: string;
}

export interface StepTimelineEntry {
  stepId: string;
  action: string;
  status: 'success' | 'failed' | 'bypassed';
  startedAt: string;
  durationMs: number;
  resolvedParams: Record<string, any>;
  output: any;
  error?: string;
}

export interface SimulationResult {
  executionId?: string;
  success?: boolean;
  status?: 'completed' | 'failed' | 'cycle_terminated';
  executedSteps?: string[];
  bypassedSteps?: string[];
  unreachedSteps?: string[];
  executedStepIds?: string[];
  bypassedStepIds?: string[];
  unreachedStepIds?: string[];
  activeEdgeIds?: string[];
  bypassedEdgeIds?: string[];
  stepOutputs: Record<string, any>;
  errors?: Record<string, any>;
  timeline?: StepTimelineEntry[];
  executionLogs?: StepExecution[];
  transitionLogs?: TransitionExecution[];
  finalContext?: ExecutionContext;
}

export interface SimulationState {
  isRunning: boolean;
  input: SimulationInput;
  initialVars: Record<string, any>;
  stepMockOverrides: Record<string, any>;
  result: SimulationResult | null;
}
